wagtail/admin/views/pages/choose_parent.py

```diff
@@ -18,6 +19,67 @@ class ChooseParentView(WagtailAdminTemplateMixin, FormView):
     index_url_name = None
     page_title = gettext_lazy("Choose parent")
 
+    def get_valid_parent_pages(self, user):
+        """
+        Identifies possible parent pages for the current user by first looking
+        at allowed_parent_page_models() on self.model to limit options to the
+        correct type of page, then checking permissions on those individual
+        pages to make sure we have permission to add a subpage to it.
+        """
+        # Get queryset of pages where this page type can be added
+        allowed_parent_page_content_types = list(
+            ContentType.objects.get_for_models(
+                *self.model.allowed_parent_page_models()
+            ).values()
+        )
+        allowed_parent_pages = Page.objects.filter(
+            content_type__in=allowed_parent_page_content_types
+        )
+
+        # Get queryset of pages where the user has permission to add subpages
+        if user.is_superuser:
+            pages_where_user_can_add = Page.objects.all()
+        else:
+            pages_where_user_can_add = Page.objects.none()
+
+            from wagtail.permission_policies.pages import PagePermissionPolicy
+
+            perms = {
+                perm
+                for perm in PagePermissionPolicy().get_cached_permissions_for_user(user)
+                if perm.permission.codename == "add_page"
+            }
+
+            for perm in perms:
+                # user has add permission on any subpage of perm.page
+                # (including perm.page itself)
+                pages_where_user_can_add |= Page.objects.descendant_of(
+                    perm.page, inclusive=True
+                )
+
+        # Combine them
+        return allowed_parent_pages & pages_where_user_can_add
+
+    def get(self, request, *args, **kwargs):
+        parents = self.get_valid_parent_pages(request.user)
+
+        # There's only one available parent for this page type for this
+        # user, so we send them along with that as the chosen parent page
+        if len(parents) == 1:
+            parent = parents[0]
+            parent_id = quote(parent.pk)
```

Comment:
One of the reasons why I didn't include this in the original PR was because of performance concerns raised in https://github.com/wagtail-nest/wagtail-modeladmin/issues/19 as I mentioned in #11734.

The original ModelAdmin implementation evaluates the whole queryset of pages and provide a form input with those pages as the options. When rendering each option, it does another query to retrieve the parents so that it can render the parent titles with a `|` separator: https://github.com/wagtail-nest/wagtail-modeladmin/blob/742fbe8d5e942ba641af4c25fb181c65ad4858ed/wagtail_modeladmin/forms.py#L6-L13

I tried to avoid the performance issue by using a chooser, which eliminates the extra queries and provides a more natural interface for navigating through the pages – but I forgot that we lost the redirect in the process.

Since we only need to make this query to do a redirect if there's only one valid parent page, perhaps we can simplify the query to either:

- Two queries: one for `.count()` and another for `.get()` if `count == 1`, e.g.

```suggestion
        parents = self.get_valid_parent_pages(request.user)

        # There's only one available parent for this page type for this
        # user, so we send them along with that as the chosen parent page
        if parents.count() == 1:
            parent_id = quote(parents.values_list("pk", flat=True).get())
```

or

- One query, pretty much like the current one, but limit it to only two results and only get the primary key of the pages.

```suggestion
        parent_ids = self.get_valid_parent_pages(request.user).values_list("pk", flat=True)[:2]

        # There's only one available parent for this page type for this
        # user, so we send them along with that as the chosen parent page
        if len(parent_ids) == 1:
            parent_id = quote(parent_ids[0])
```

Reply:
@laymonage thank you, that totally makes sense performance wise. Its updated!